src/board/BoardLogic.java

```diff
@@ -289,23 +297,66 @@ protected boolean removeApple(int x, int y) {
      * place apples on the board until MAX_APPLES_ON_BOARD value is reached
      */
     protected void setApples() {
+        // get all valid fields
+        List<Field> validFields = new ArrayList<>(getValidAppleFields());
 
         // add apples until value of MAX_APPLES_ON_BOARD is reached
-        while (apples.size() < MAX_APPLES_ON_BOARD) {
-
-            // get random field with field status empty
-            Field appleField;
-            do {
-                appleField = getRandomField();
-
-            } while (!(fields[appleField.getPosX()][appleField.getPosY()].getState() == FieldState.Empty));
+        while (apples.size() < MAX_APPLES_ON_BOARD && validFields.size() > 0) {
+            // get random valid field
+            Field appleField = validFields.get((int) (Math.random() * validFields.size()));
+            validFields.remove(appleField);
 
             // set field state of random field to value apple and add apple to apple-list
-            fields[appleField.getPosX()][appleField.getPosY()].setState(FieldState.Apple);
+            appleField.setState(FieldState.Apple);
             apples.add(appleField);
         }
     }
 
+    /**
+     * generate a set of all valid apple fields
+     *
+     * @return a set containing all fields where an apple could be reached by a snake
+     */
+    protected Set<Field> getValidAppleFields() {
+        // checking from all snake heads
+        Deque<Field> checkFields = new ArrayDeque<>();
+        for (LinkedList<Field> snake : snakesLocation) {
+            checkFields.add(snake.getLast());
+        }
+
+        Set<Field> validFields = new HashSet<Field>();
+        while (!checkFields.isEmpty()) {
+            Field field = checkFields.removeLast();
+            Field tempField;
+            // checking in all 4 directions
+            if (field.getPosY() > 0
+                && !validFields.contains(tempField = getFields()[field.getPosX()][field.getPosY() - 1])
```

Comment:
Man könnte statt in allen validFields zu suchen vllt folgendes machen:
Am Anfang der Funktion erstellt man ein 2-dim boolean Array, dass so groß ist, wie das originale fields array und alles auf false setzten. false bedeutet dann, dass man noch nicht bei diesem Feld war und true, dass man das schon durchsucht hat.
Dann würde man schreiben `(!visitFields[field.getPosX()][field.getPosY() - 1])` und in den ifs schreiben `visitFields[field.getPosX()][field.getPosY() - 1] = true`.
Man muss natürlich am Anfang noch die Positionen der Startpunkte (Schlangenköpfe) auf true setzten.

So läuft die Berechnung etwas schneller denke ich.

Reply:
Sollte keinen Unterschied machen meine ich, habe hier bewusst HashSet verwendet um diese Checks sehr schnell zu machen, Komplexität ist auch O(1).

Reply:
Ahh ja das hab ich übersehen. Passt!